core/os/thread.h

```diff
@@ -74,26 +79,25 @@ class Thread {
 private:
 	friend class Main;
 
-	static ID main_thread_id;
-
-	static uint64_t _thread_id_hash(const std::thread::id &p_t);
-
-	ID id = _thread_id_hash(std::thread::id());
+	ID id = UNASSIGNED_ID;
+	static SafeNumeric<uint64_t> id_counter;
 	static thread_local ID caller_id;
 	std::thread thread;
 
 	static void callback(ID p_caller_id, const Settings &p_settings, Thread::Callback p_callback, void *p_userdata);
 
 	static PlatformFunctions platform_functions;
 
+	static void make_main_thread() { caller_id = MAIN_ID; }
+
 public:
 	static void _set_platform_functions(const PlatformFunctions &p_functions);
 
 	_FORCE_INLINE_ ID get_id() const { return id; }
 	// get the ID of the caller thread
 	_FORCE_INLINE_ static ID get_caller_id() { return caller_id; }
```

Comment:
```suggestion
	_FORCE_INLINE_ static ID get_caller_id() {
		if (unlikely(caller_id == UNASSIGNED_ID)) {
			caller_id = ...;
		}
		return caller_id;
	}
```

As a possible alternative to `attach_external_thread`.

Reply:
I thought about this, but I am not sure what are the performance implications. It may be negligible in our case though.

Reply:
On second thought, this makes a lot more sense, so added it.